tempesta_fw/t/unit/tfw_fuzzer.c

```diff
@@ -0,0 +1,563 @@
+				gen_vector[TRANSFER_ENCODING].max);
+	}
+
+	return v;
+}
+
+static int add_body(char *s1, int type)
+{
+	size_t len = strlen(s1), cont_len,
+	       i, body_size = gen_vector[BODY_SIZE].i * 256;
+	char *rc;
+	int err;
+
+	for (i = 0; i < body_size; ++i)
+		s1[len + i] = a_body[((i + gen_vector[CONTENT_LENGTH].i +
+				       gen_vector[BODY_SIZE].i + len) ^
+				       (a_body[i] + len)) % sizeof(a_body)];
+
+	err = kstrtoul(content_length[gen_vector[CONTENT_LENGTH].i].s,
+			10, &cont_len);
+	if (err)
+		return FUZZ_INVALID;
+	if (cont_len > body_size)
+		return FUZZ_INVALID;
+	if (type == FUZZ_REQ)
+		return FUZZ_VALID;
+
+	rc = resp_code[gen_vector[RESP_CODE].i].s;
+	if (!strstr(rc, "100") || !strstr(rc, "101") || !strstr(rc, "102") ||
+			!strstr(rc, "204") || !strstr(rc, "304"))
+		return FUZZ_INVALID;
+
+	return FUZZ_VALID;
+}
+
+static int add_duplicates(char *s1, int n, int(*add_func)(char *s1))
+{
+	int i, tmp, v = 0;
+
+	for (i = 0; i < gen_vector[DUPLICATES].i; ++i) {
+		tmp = gen_vector[n].i;
+
+		gen_vector[n].i = (i + gen_vector[n].i +
+				   gen_vector[SPACES].i) %
+				  gen_vector[n].max;
+		v |= (*add_func)(s1);
+		strcat(s1, "\r\n");
+		gen_vector[n].i = tmp;
+	}
+
+	if (gen_vector[n].singular && i > 0)
+		return FUZZ_INVALID;
+	return v;
+}
+
+/* Returns:
+ * FUZZ_VALID if the result is a valid request,
+ * FUZZ_INVALID if it's invalid,
+ * FUZZ_END if the request sequence is over.
+ * `move` is how many gen_vector's elements should be changed
+ * each time a new request is generated, should be >= 1. */
+int fuzz_gen(char *str, int move, int type)
+{
+	int i, ret, v = 0;
+
+	if (str == NULL)
+		return -EINVAL;
+	if (move < 1)
+		return -EINVAL;
+	*str = '\0';
+
+	if (type == FUZZ_REQ) {
+		v |= add_method(str);
+		add_spaces(str);
+
+		add_uri_path_start(str);
+		if (!gen_vector[URI_PATH_DEPTH].i)
+			addch(str, '/');
+		for (i = 0; i < gen_vector[URI_PATH_DEPTH].i; ++i) {
+			addch(str, '/');
+			v |= add_uri_file(str);
+		}
+		add_spaces(str);
+	}
+
+	strcat(str, "HTTP/");
+	v |= add_http_ver(str);
+	if (type == FUZZ_RESP) {
+		add_spaces(str);
+		v |= add_resp_code(str);
+	}
+	strcat(str, "\r\n");
+
+	v |= add_connection(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, CONNECTION, &add_connection);
+
+	v |= add_user_agent(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, USER_AGENT, &add_user_agent);
+
+	v |= add_host(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, HOST, &add_host);
+
+	v |= add_x_forwarded_for(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, X_FORWARDED_FOR, &add_x_forwarded_for);
+
+	v |= add_content_type(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, CONTENT_TYPE, &add_content_type);
+
+	v |= add_content_length(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, CONTENT_LENGTH, &add_content_length);
+
+	v |= add_transfer_encoding(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, TRANSFER_ENCODING, &add_transfer_encoding);
```

Comment:
Not enough headers. Here is an example of standard browser request:

         GET / HTTP/1.1
         Host: 172.16.0.5
         User-Agent: Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:38.0) Gecko/20100101 Firefox/38.0
         Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
         Accept-Language: en-US,en;q=0.5
         Accept-Encoding: gzip, deflate
         Connection: keep-alive

So we should at least test these Accept-* headers. Also Cookie and Set-Cookie are missed.